File: Wallet-program/src/main/java/org/Action.java
package org;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class Action implements Serializable {
    private static final long serialVersionUID = 1L;
    private Map<String, WalletModel> wallet;
    private static final String FILENAME = "wallet.ser";
    public Action() {
        wallet = new HashMap<>();
        chargeData();
    }

    private void chargeData() {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(FILENAME)))) {
            Object obj = ois.readObject();
            if (obj instanceof Map) {
                wallet = (Map<String, WalletModel>) obj;
            }
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Impossible de charger les données existantes.");
        }
    }

    private void saveData() {
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(FILENAME)))) {
            oos.writeObject(wallet);
        } catch (IOException e) {
            System.out.println("Erreur lors de la sauvegarde des données.");
        }
    }

    public void createWallet(String name, int len_bill, int len_card, STATUS status) {
        wallet.put(name, new WalletModel(len_card, len_bill, status));
        saveData();
        System.out.println("Portefeuille créé pour " + name);
    }

    public void addSource(String name,String source, double value) {
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            portefeuille.addSource(name, source, value);
            saveData();
        } else {
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

    public void expenses(String name, String source, double value) {
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            portefeuille.addExpense(name,source,value);
        } else {
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

    public void resetWallet(String name){
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            portefeuille.setAmount(name);
            saveData();
            System.out.println("Portefeuille réinitialisé pour " + name);
        } else {
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

    public void showListExpenses(String name){
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            portefeuille.getListExpense(name);
        }
        else{
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

    public void showListSource(String name){
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            portefeuille.getListSource(name);
        }
        else{
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

    public void showWalletModel(String name) {
        WalletModel portefeuille = wallet.get(name);
        if (portefeuille != null) {
            System.out.println("Votre compte est de : "+ portefeuille.getAmountsForWallet(name));
        }
        else{
            System.out.println("Portefeuille introuvable pour " + name);
        }
    }

}

File: WalletProgramPy/action.py
import pickle

from model import WalletModel


class Action:
    def __init__(self):
        self.wallet = {}
        self.FILENAME = "wallet.pkl"
        self.charge_data()

    def charge_data(self):
        try:
            with open(self.FILENAME, 'rb') as file:
                obj = pickle.load(file)
                if isinstance(obj, dict):
                    self.wallet = obj
        except (IOError, pickle.UnpicklingError) as e:
            print("Impossible de charger les données existantes.")

    def save_data(self):
        try:
            with open(self.FILENAME, 'wb') as file:
                pickle.dump(self.wallet, file)
        except IOError as e:
            print("Erreur lors de la sauvegarde des données.")

    def create_wallet(self, name):
        self.wallet[name] = WalletModel()
        self.save_data()
        print(f"Portefeuille créé pour {name}")

    def add_source(self, name, source, value):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            wallet_model.add_source(name, source, value)
            self.save_data()
        else:
            print(f"Portefeuille introuvable pour {name}")

    def expenses(self, name, source, value):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            wallet_model.add_expense(name, source, value)
        else:
            print(f"Portefeuille introuvable pour {name}")

    def reset_wallet(self, name):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            wallet_model.set_amount(name)
            self.save_data()
            print(f"Portefeuille réinitialisé pour {name}")
        else:
            print(f"Portefeuille introuvable pour {name}")

    def show_list_expenses(self, name):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            wallet_model.get_list_expense(name)
        else:
            print(f"Portefeuille introuvable pour {name}")

    def show_list_source(self, name):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            wallet_model.get_list_source(name)
        else:
            print(f"Portefeuille introuvable pour {name}")

    def show_wallet_model(self, name):
        wallet_model = self.wallet.get(name)
        if wallet_model:
            print(f"Votre compte est de : {wallet_model.get_amounts_for_wallet(name)}")
        else:
            print(f"Portefeuille introuvable pour {name}")
